lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs

```diff
@@ -87,6 +92,24 @@ data CoinSelectionOptions e = CoinSelectionOptions
             -- ^ Returns any backend-specific error regarding coin selection
     } deriving (Generic)
 
+-- | Calculate the sum of all input values
+inputBalance :: CoinSelection -> Word64
+inputBalance =  foldl' (\total -> addTxOut total . snd) 0 . inputs
+
+-- | Calculate the sum of all output values
+outputBalance :: CoinSelection -> Word64
+outputBalance = foldl' addTxOut 0 . outputs
+
+-- | Calculate the sum of all output values
+changeBalance :: CoinSelection -> Word64
+changeBalance = foldl' addCoin 0 . change
```

Comment:
Would it not be simpler to define `Semigroup` and `Monoid` instances for `Coin`?

Then we can write these functions as so:

```hs
-- | Calculate the sum of all input values                                                                                                                                            
inputBalance :: CoinSelection -> Coin                                                                                                                                                 
inputBalance = mconcat . fmap (coin . snd) . inputs                                                                                                                                   
                                                                                                                                                                                      
-- | Calculate the sum of all output values                                                                                                                                           
outputBalance :: CoinSelection -> Coin                                                                                                                                                
outputBalance = mconcat . fmap coin . outputs                                                                                                                                         
                                                                                                                                                                                      
-- | Calculate the sum of all output values                                                                                                                                           
changeBalance :: CoinSelection -> Coin                                                                                                                                                
changeBalance = mconcat . change     
```

(`addTxOut` and `addCoin` become redundant.)

Reply:
The main why we don't have a Monoid / Semigroup instance here is the same reason we don't have Monoid / Semigroup instance on numbers:

A Monoid / Semigroup is a relationship between a data-type and a binary operation. So in the end, it doesn't make much sense to talk about `Int` as a Monoid, but rather, one will talk about `(Int, +)` or `(Int, *)`. As a matter of fact, there can be multiple valid binary operations for making Monoid instances, and one would instead define `Monoid` / `Semigroup` instances on data-types which also captures what is this binary operation. e.g.:

- [Product](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Monoid.html#t:Product)
- [Sum](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Monoid.html#t:Sum)
- [Last](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Monoid.html#t:Last)
- [First](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Semigroup.html#t:First)
- [Max](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Semigroup.html#t:Max)
- [Min](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Semigroup.html#t:Min)

The same rationale applies to `Coin`. And to this, I'll also add that, even if it's somewhat useful in some computations, a `Coin 0` (with would be the typical choice for `mempty`) doesn't make much sense. Conceptually, there's no such thing as an empty coin!

Reply:
So, if `Coin` had an instance of `Num`, we could simply write:

```hs
changeBalance = getSum . mconcat . change
```

I don't think we have a `Num` instance for "safety" reasons, but it'd be less controversial than a `Monoid` or `Semigroup` instance :)

Reply:
> I don't think we have a Num instance for "safety" reasons.

What are the reasons?

> Conceptually, there's no such thing as an empty coin.

Well, although we don't have empty coins on the blockchain, we _do_ start with an empty value of `0` in the functions `inputBalance`, `outputBalance` and `changeBalance`.

What is the conceptual type of this `0`, if not "coin value", or something similar?

The issue with returning `Word64` from these functions is that we lose some information about the types of values that can be produced.

Reply:
A `Num` instance on `Coin` is unsafe because many operations in this instance can't be defined for `Coin`. Half of the functions would be partial whereas the only thing we usually want is summing coins.

As for the balance calculation, we do start with a `0 :: Word64` which is not an empty Coin ;).. We are already in the output domain. Really, Coins are bank notes. You don't have notes of `0 EUR` although your account balance can be of `0 EUR` (or even negative...). They are not the same thing.

Reply:
Also, we're not loosing any information here because, as soon as one talks about balance, one does not talks about coins anymore. Rather, one talks about their value.

Reply:
> Also, we're not loosing any information here

My point is that there's nothing to distinguish the type of output values from other values that happen to be `Word64`. The function could be more self-documenting (at both the point of definition and the point of use) if the return type was something particular, rather than just `Word64`. 

I note that in other places we use `Quantity "lovelace" Word64`. Perhaps we could use something like that here?

Reply:
My reference to "information loss" refers to the fact that within the functions `changeBalance`, `inputBalance` and `outputBalance`, we _know_ we're operating on something that is a _balance_ (or a _quantity_) of Lovelace. But by the time we've converted it to a generic `Word6` and it's been passed around by the caller to somewhere else, that information is no longer present in the type system. It's just a `Word64.

Reply:
Ho yes, but you're touching somethk G different. Returning a `Quantity "lovelace" Word64` or another adhoc type `Balance` would be best. But that's a completely different question.